File: Python/Day2/2.py

safe = 0

def validate_arr(arr) -> bool:
    prev = arr[0]
    check = None
    for i in arr[1:]:
        if check is None:
            if prev < i:
                check = "increasing"
            if prev > i:
                check = "decreasing"
        if abs(i - prev) > 3 or abs(i - prev) < 1:
            return False
        if check == "increasing" and not prev < i:
            return False
        if check == "decreasing" and not prev > i:
            return False
        prev = i
    return True

with open("./Input/2.txt") as f:
    lines = f.readlines()
    for line in lines:
        split = line.split(" ")
        arr = [int(x) for x in split]
        result = validate_arr(arr)
        if not result:
            for i in range(len(arr)):
                result = validate_arr(arr[:i] + arr[i + 1:])
                if result:
                    break
        if result:
            safe += 1
    print(safe)
        

File: Go/Day2/2.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	data, _ := os.Open("./Input/2.txt")

	scanner := bufio.NewScanner(data)
	total := 0

	for scanner.Scan() {
		arr := line_to_arr(scanner.Text())
		fmt.Println(arr)
		result := eval_line(arr)
		fmt.Println(result)
		if !result {
			for i := range arr {
				new_arr := skip_index(i, arr)
				result = eval_line(new_arr)
				if result {
					break
				}
			}
		}

		if result {
			total += 1
		}
	}

	fmt.Println(total)
}

func skip_index(x int, arr []int) []int {
	new_arr := make([]int, len(arr)-1)
	count := 0
	for i, v := range arr {
		if i == x {
			continue
		}
		new_arr[count] = v
		count++
	}

	return new_arr
}

func eval_line(arr []int) bool {
	var eval func(prev int, curr int) bool
	var prev int
	for i, v := range arr {
		if i == 0 {
			prev = v
			continue
		}
		if eval == nil {
			eval = set_eval(prev, v)
		}
		if int(math.Abs(float64(prev-v))) > 3 || int(math.Abs(float64(prev-v))) < 1 {
			return false
		}
		if !eval(prev, v) {
			return false
		}
		prev = v
	}

	return true
}

func set_eval(prev int, curr int) func(int, int) bool {
	if prev < curr {
		return func(prev int, curr int) bool {
			return prev < curr
		}
	} else {
		return func(prev, curr int) bool {
			return prev > curr
		}
	}
}

func line_to_arr(line string) []int {
	arr_str := strings.Split(line, " ")
	arr := make([]int, len(arr_str))
	for i, v := range arr_str {
		num, _ := strconv.Atoi(v)
		arr[i] = num
	}

	return arr
}
